### PR TITLE
test for max stack size

### PR DESCRIPTION
**Pull request by mubarak23**
- [x] issue #191 
- [x] follows contribution [guide](https://github.com/keep-starknet-strange/shinigami/blob/main/CONTRIBUTING.md)
- [x] code change includes tests

### CODE REVIEW

File: src/opcodes/tests/test_stack.cairo [hunk 1]
```diff
@@ -291,6 +291,17 @@ fn test_op_2rot_insufficient_items() {
...
... OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_1 OP_2ROT";
```
Comment (b-j-roberts) [inline src/opcodes/tests/test_stack.cairo:297]: Could you change this to use a for loop instead to create the string. You should be able to do something like 
```
let mut program = "";
let op_1_string = "OP_1 ";
// loop
program.append(op_1_string)
```

File: src/opcodes/tests/test_stack.cairo [hunk 2]
```diff
@@ -291,6 +291,17 @@ fn test_op_2rot_insufficient_items() {
...
+
```
Comment (b-j-roberts) [inline src/opcodes/tests/test_stack.cairo:298]: Is the `OP_2ROT` needed at the end?

File: src/opcodes/tests/test_stack.cairo [hunk 3]
```diff
@@ -291,6 +291,17 @@ fn test_op_2rot_insufficient_items() {
     utils::check_dstack_size(ref engine, 5);
 }
 
+#[test]
+fn test_max_stack_size() {
```
Comment (b-j-roberts) [inline src/opcodes/tests/test_stack.cairo:295]: Could you add a test that adds 1000 elements and doesn't throw an error?

**Review by b-j-roberts — CHANGES_REQUESTED** [review]

File: src/opcodes/tests/test_stack.cairo (hunk 3 above)
Comment (mubarak23) [reply src/opcodes/tests/test_stack.cairo:295]: ok

File: src/opcodes/tests/test_stack.cairo (hunk 2 above)
Comment (mubarak23) [reply src/opcodes/tests/test_stack.cairo:298]: to complete the execution script

File: src/opcodes/tests/test_stack.cairo (hunk 1 above)
Comment (mubarak23) [reply src/opcodes/tests/test_stack.cairo:297]: can we apppend a value to ByteArray

Comment (mubarak23) [reply src/opcodes/tests/test_stack.cairo:297]: we cannot append a value to string, as append method only works if it an array

**Comment by mubarak23** [comment]
@b-j-roberts This PR is Ready for review

File: src/opcodes/tests/test_stack.cairo [hunk 4]
```diff
@@ -291,6 +291,36 @@ fn test_op_2rot_insufficient_items() {
     utils::check_dstack_size(ref engine, 5);
 }
 
+#[test]
+fn test_max_stack_size_more() {
+    let mut program: ByteArray = "";
+    let op_1_string = "OP_1 ";
+    let mut index: u64 = 0;
+    while index < 1045 {
```
Comment (b-j-roberts) [inline src/opcodes/tests/test_stack.cairo:299]: Is there a reason to test with 1045 elements? I think 2 tests would be preffered, one adding 1000 elements that doesn't fail and one adding 1001 that does fail.

**Review by b-j-roberts — CHANGES_REQUESTED** [review]

File: src/opcodes/tests/test_stack.cairo (hunk 4 above)
Comment (mubarak23) [reply src/opcodes/tests/test_stack.cairo:299]: no reason, i will make 1001

File: src/opcodes/tests/test_stack.cairo [hunk 5]
```diff
@@ -291,6 +291,36 @@ fn test_op_2rot_insufficient_items() {
...
+        index += 1;
+    };
+
+    let mut engine = utils::test_compile_and_run_err(program, Error::SCRIPT_STACK_SIZE_EXCEEDED);
+
+    utils::check_dstack_size(ref engine, 1001);
+}
+
+#[test]
+fn test_max_stack_size_equal() {
```
Comment (b-j-roberts) [inline src/opcodes/tests/test_stack.cairo:310]: Both tests are adding 1001 elements at the moment

**Review by b-j-roberts — CHANGES_REQUESTED** [review]

File: src/opcodes/tests/test_stack.cairo (hunk 5 above)
Comment (mubarak23) [reply src/opcodes/tests/test_stack.cairo:310]: hmm, i see what is happening here

File: src/opcodes/tests/test_stack.cairo (hunk 4 above)
Comment (mubarak23) [reply src/opcodes/tests/test_stack.cairo]: > Is there a reason to test with 1045 elements? I think 2 tests would be preffered, one adding 1000 elements that doesn't fail and one adding 1001 that does fail.

am confuse here, 

can you explain further please

**Comment by mubarak23** [comment]
this test function throw the following error 

```
#[test]
fn test_max_stack_size_more() {
    let mut program: ByteArray = "";
    let op_1_string = "OP_1 ";
    let mut index: u64 = 0;
    while index < 1000 {
        program.append(@op_1_string);
        index += 1;
    };

    let mut engine = utils::test_compile_and_run_err(program, Error::SCRIPT_STACK_SIZE_EXCEEDED);

    utils::check_dstack_size(ref engine, 1000);
}
```

this errors 

```

failures:
   shinigami::opcodes::tests::test_stack::test_max_stack_size_more - Panicked with "Execution of the program did not fail as expected".
```
which mean putting 1000 item in the stack does not break it  @b-j-roberts

**Comment by b-j-roberts** [comment]
> which mean putting 1000 item in the stack does not break it @b-j-roberts

Yes, that is the expected behavior. 1000 items on the stack should work fine. So you can use `utils::test_compile_and_run` instead of the error one.

**Pull request merged**